nvflare/edge/assessors/model_update.py

```diff
@@ -54,44 +37,68 @@ def __init__(
+        For asynchronous updates, the staleness is calculated based on the starting and current model version.
+        And the aggregation scheme and weights are calculated following FedBuff paper "Federated Learning with Buffered Asynchronous Aggregation".
+
+        Args:
+            persistor_id (str): ID of the persistor component used to load and save models.
+            num_updates_for_model (int): Number of updates required before generating a new model version.
+            max_model_version (int): Maximum number of model versions allowed in the workflow.
+            max_model_history (int): Maximum number of historical model versions to keep in memory.
+            device_selection_size (int): Number of devices to select for each model update round.
+            min_hole_to_fill (int, optional): Minimum number of empty slots in device selection before refilling. Defaults to 1 - once received an update, immediately sample a new device and send the current task to it.
+            global_lr (float, optional): Global learning rate for model aggregation. Defaults to 1.0.
+            staleness_weight (bool, optional): Whether to apply staleness weighting to model updates. Defaults to False - no staleness weighting.
+            device_reuse (bool, optional): Whether to allow reusing devices across different model versions. Defaults to True.
+            const_selection (bool, optional): Whether to use constant device selection across rounds. Defaults to False.
         """
         Assessor.__init__(self)
+        # Claim as self attributes for JobAPI
         self.persistor_id = persistor_id
         self.persistor = None
-        self.current_model = None
-        self.current_model_version = 0
-        self.current_selection = {}
-        self.current_selection_version = 0
-        self.updates = {}  # model_version => _ModelState
-        self.available_devices = {}
-        self.used_devices = {}
         self.num_updates_for_model = num_updates_for_model
         self.max_model_version = max_model_version
         self.max_model_history = max_model_history
         self.device_selection_size = device_selection_size
         self.min_hole_to_fill = min_hole_to_fill
         self.device_reuse = device_reuse
         self.const_selection = const_selection
+        self.global_lr = global_lr
+        self.staleness_weight = staleness_weight
+
         self.update_lock = threading.Lock()
         self.start_time = None
         self.register_event_handler(EventType.START_RUN, self._handle_start_run)
+        self.model_manager = ModelManager(
+            num_updates_for_model=num_updates_for_model,
+            max_model_version=max_model_version,
+            max_model_history=max_model_history,
+            global_lr=global_lr,
+            staleness_weight=staleness_weight,
+        )
+        self.device_manager = DeviceManager(
```

Comment:
Do not hard-code to a particular impl of DeviceManager. Instead, you should get the device manager from the specified component id.

Reply:
done